src/Microsoft.AspNet.OData.Shared/Formatter/Deserialization/DeserializationHelpers.cs

```diff
@@ -19,7 +23,54 @@ internal static class DeserializationHelpers
         internal static void ApplyProperty(ODataProperty property, IEdmStructuredTypeReference resourceType, object resource,
             ODataDeserializerProvider deserializerProvider, ODataDeserializerContext readContext)
         {
-            IEdmProperty edmProperty = resourceType.FindProperty(property.Name);
+            WebApiRequestMessage request = null;
+            if(readContext != null)
+            {
+                request = (WebApiRequestMessage)readContext.InternalRequest;
+            }
+
+            bool enableCaseInsensitiveModelBinding = false;
+
+            if(request != null)
+            {
+#if NETCORE
+                // In AspNetCore, we add our setting by adding a service to the ServiceCollection as below:
+                // services.TryAddEnumerable(ServiceDescriptor.Singleton<IODataModelBinderSettings>(new ODataModelBinderSettings { EnableCaseInsensitiveModelBinding = true }));
+                // We need to resolve the service so that we can get our setting.
+                var settings = request.innerRequest.HttpContext.RequestServices.GetService<IODataModelBinderSettings>();
+                if(settings != null)
+                {
+                    enableCaseInsensitiveModelBinding = settings.EnableCaseInsensitiveModelBinding;
+                }
+#else
+                // In AspNet classic, we add our setting to HttpConfiguration.Properties as below:
+                // configuration.Properties[EnableCaseInsensitiveModelBindingKey] = true;
+                var container = request.innerRequest.GetConfiguration();
+                if(container != null)
+                {
+                    enableCaseInsensitiveModelBinding = container.HasEnabledCaseInsensitiveModelBinding();
+                }
+#endif
+            }
+
+            IEdmProperty edmProperty;
+
+            if (enableCaseInsensitiveModelBinding)
+            {
+                IDictionary<string, string> propertyNamesMapping = new Dictionary<string, string>();
+                var sProperties = resourceType.StructuralProperties();
+                foreach (var prop in sProperties)
+                {
+                    propertyNamesMapping.Add(prop.Name.ToUpperInvariant(), prop.Name);
+                }
```

Comment:
What if the model has two different properties that vary only by case?  Can we first try a case-sensitive match and, if that fails, fall back to case-insensitive (note that we typically fail if there were two case-insensitive matches and no case sensitive matches).

Reply:
#Resolved